distributed/shuffle/tests/test_rechunk.py

```diff
@@ -1141,3 +1143,69 @@ def test_split_axes_with_zero():
         [[Split(0, 0, slice(0, 1, None)), Split(1, 0, slice(1, 2, None))]],
     ]
     assert result == expected
+
+
+@gen_cluster(client=True)
+async def test_preserve_writeable_flag(c, s, a, b):
+    """Make sure that the shuffled array doesn't accidentally become read-only after
+    the round-trip to e.g. read-only file descriptors or byte objects as buffers
+    """
+    arr = da.random.random(10, chunks=5)
+    arr = arr.rechunk(((4, 6),), method="p2p")
+    arr = arr.map_blocks(lambda chunk: chunk.flags["WRITEABLE"])
+    out = await c.compute(arr)
+    assert out.tolist() == [True, True]
+
+
+def get_host_array(a):
+    assert isinstance(a, np.ndarray)
+    while True:
+        if isinstance(a, memoryview):
+            a = a.obj
+        elif isinstance(a, np.ndarray):
+            if a.base:
+                a = a.base
+            else:
+                return a
+        else:
+            # distributed.comm.utils.host_array() uses numpy.empty()
+            raise AssertionError("unreachable")  # pragma: nocover
+
+
+@gen_cluster(client=True, config={"distributed.p2p.disk": False})
+async def test_rechunk_in_memory_shards_dont_share_buffer(c, s, a, b):
+    """Test that, if two shards are sent in the same RPC call and they contribute to
+    different output chunks, downstream tasks don't need to consume all output chunks in
+    order to release the memory of the output chunks that have already been consumed.
+
+    This can happen if all numpy arrays in the same RPC call share the same buffer
+    coming out of the TCP stack.
+    """
+    in_map = Event()
+    block_map = Event()
+
+    def blocked(chunk, in_map, block_map):
+        in_map.set()
+        block_map.wait()
+        return chunk
+
+    arr = da.random.random((10, 10), chunks=(-1, 1))
+    arr = arr.rechunk((1, -1), method="p2p")
+    arr = arr.map_blocks(blocked, in_map=in_map, block_map=block_map, dtype=arr.dtype)
+    fut = c.compute(arr)
+    await in_map.wait()
+
+    [run] = a.extensions["shuffle"].shuffle_runs._runs
+    shards = [
+        s3 for s1 in run._disk_buffer._shards.values() for s2 in s1 for _, s3 in s2
+    ]
```

Comment:
@hendrikmakait this is ugly; could you think of a nicer way to get the shards in the MemoryBuffer?